### PR TITLE
Android: Fixes #3872: Allow background syncing during the first 5 minutes of the sync for outgoing changes

### LINKED ISSUE #3872: background sync on mobile devices
There are various issues ([1](https://github.com/laurent22/joplin/issues/727), [2](https://github.com/laurent22/joplin/issues/1029), [3](https://github.com/laurent22/joplin/issues/1501), [4](https://github.com/laurent22/joplin/issues/1645), [5](https://github.com/laurent22/joplin/issues/3671), …) and threads ([1](https://discourse.joplinapp.org/t/android-background-sync-interval/1746), [2](https://discourse.joplinapp.org/t/automatic-sync-on-phone-app/3056)) on it but it's still not clear to me what exactly isn't technically possible to implement regarding this.

[This comment from September 2018 says](https://github.com/laurent22/joplin/issues/727#issuecomment-418877957):

> Cannot currently be fixed due a limitation of the underlying framework

Which might be true but that's what libraries are for, right?

And [this comment from October 19](https://discourse.joplinapp.org/t/automatic-sync-on-phone-app/3056/8) claims:

> we rely on existing libs, which for now don’t exist on React Native

But it doesn't mention _which_ exact libraries would be needed. The logic for syncing is obviously already there. So in theory only the support for doing exactly this sync in a background task would be needed.
And there are libraries for react-native that are doing exactly this since a few years. For example
* [`react-native-background-fetch`](https://github.com/transistorsoft/react-native-background-fetch) for Android & iOS (first version from August 2016)
* [`react-native-sync-adapter`](https://github.com/ferrannp/react-native-sync-adapter) using the Android SyncAdapter (first version from April 2017)
* [`react-native-background-task`](https://github.com/jamesisaac/react-native-background-task) again for Android & iOS (first version from February 2017)
* [`react-native-background-job`](https://github.com/vikeri/react-native-background-job) for Android (available since December 2016)

Also there's [this comment from June 2019](https://github.com/laurent22/joplin/issues/1645#issuecomment-500384473)

> Most likely not possible. Android phones these days have services that kill off apps that aren't used, whether you want to keep them running or not. https://dontkillmyapp.com/

Well,

1. it only affects certain vendors in an extreme way. Not every phone is affected.
2. The exact same website you mentioned shows some workarounds for this issue.
3. Apps are usually killed after a while — not immediately. So Joplin could probably still sync for a while until it would be killed.
4. Some FOSS apps which are showing notifications while they are running doesn't seem to be affected (or at least doesn't be to seem affected so often). For instance [DNS66](https://github.com/julian-klode/dns66) or [personalDNSfilter](https://github.com/IngoZenz/personaldnsfilter).

I would obviously understand if the issue would only be the lack of people working on this. In that case this issue could be kept open for tracking, avoiding duplicates and having the possibility to show this issue for possible contributors, so that they know what's still open.

But currently the given answers are at least somehow lacking some information (or not relevant / true).

Thus it would be great if someone could clarify what's really missing here.

**PS:** Thank you for your work!

### PR DESCRIPTION
**Pull request by mrjo118**
### Background

The PR implements the ability on Android for the sync to continue for a period, when the app is sent to the background or the screen is turned off. This effectively gives the sync a grace period to continue syncing, instead of being immediately suspended when the app is sent to the background, or the screen is turned off. To be clear, this solution does not run any kind of periodic sync while the app is in the background, but purely provides the ability for the sync to 'overrun' whenever Joplin is backgrounded or the screen is turned off, after a sync had been triggered. To be honest, periodic background sync does not provide a benefit anyway, if the sync is allowed to overrun and complete.

The intention of this change is to ensure outgoing changes (as part of incremental sync in daily usage) will be synced in the majority of cases, even when the user switches app or turns the screen off prematurely, before the sync has completed. This will help to prevent sync conflicts and ensure data is more readily available if the relevant device is temporarily inaccessible.

### Implementation

The solution is implemented using the react-native-background-actions library:
- While the library includes an implementation for both Android and iOS, unfortunately the iOS implementation does not seem to be fit for purpose when combined with the sync system in Joplin, so this PR implements background sync for Android only ([see comment](https://github.com/laurent22/joplin/pull/15214#issuecomment-4348136525))
- On Android, this utilises a **foreground service**, which will run whenever the sync is in progress within Joplin, and allows the sync to continue while the app is sent to the background or the screen is turned off. This requires a notification to run reliably, and notification access is requested upon triggering the sync, on Android versions which require a permission grant. On some versions of Android, this can run indefinately when battery optimisations have been disabled for the app, but on Android 16+, the running of the foreground service is limited to up to 6 hours per 24 hour period, after which the OS will kill the service and prevent it from running again within that period
- The period for background syncing has been limited to 5 minutes, due to issues if the foreground service runs for a long time ([see comment](https://github.com/laurent22/joplin/pull/15214#issuecomment-4355242120)). For incremental sync for the users day to day usage, this should allow plenty of time for the sync to completed, even with throttling due to battery optimisations (see the Testing section for performance metrics). But due to the issues with long running services and poorer sync performance, this is not suitable for the initial full sync. For a simpler implementation, the 5 minute allowance starts from the moment the sync starts, and if the user backgrounds the app after that point, or if the app is already in the background, then the foreground service will stop. Note that only the service will stop, so the sync itself will continue where it left off (without cancelling) when the app is returned to the foreground, as per the existing behaviour without background sync implemented
- In addition to the background sync ceasing after a maximum of 5 minutes, in addition, the foreground service will stop when the delta step runs, if it is a full sync. This is because it is possible for the delta step to freeze all JS execution while in the background, preventing the foreground service from ending after the 5 minute cap. The main need for background syncing is so that outgoing changes are uploaded shortly after changing a note, so therefore it is not essential to support background sync during the data step. After the foreground service has stopped, the sync will resume where it left off when the app is sent back to the foreground, providing that the app was not killed manually or by the OS
- In places where the sync is triggered via 'waitForSyncFinishedThenSync', the sync will not run using the foreground service. This is because BackgroundService.start from the library does not return a result, so the sync is run without the service so that it can be awaited. On mobile, the only place where this applies is in the ShareNoteDialog.copyLinkButton function

I have done some extensive testing on a few different Android versions, and the solution works well for what it is intended to achieve.

This partially fixes https://github.com/laurent22/joplin/issues/3872

### Considerations

- If internet access is lost during background sync, or the sync fails for any other reason, the sync will not continue until the user brings the app back to the foreground. This is an edge case
- I did a test where I change a note and immediately send the app to the background (using a keyboard shortcut on the emulator, and also using a real device), and I found that on some devices the sync does trigger (meaning they have a brief grace period for execution before the execution is paused), but on others like an Android 10 device, it did not. But all this usually means is the sync will not run in the background if the user backgrounds the app in less than one second after making a change, which is an edge case. If however a large operation is made (such as duplicating a large amount of notes), and the app is backgrounded before this operation has completed, the sync will not run in the background in this case also

### Testing

- Confirmed the sync is unaffected and works without issues on the desktop app
- Tested the change on Android 12, Android 14 and Android 16 using emulators. Also tested on a real device, which is a Samsung Galaxy S23 Ultra on Android 14, and I also did some basic tests on a Samsung Galaxy S9+ on Android 10. On all of these devices, when notification access is granted and battery optimisations are disabled, the behaviour works consistently: a notification displays whenever the sync runs (this can be tapped to open the app), and this remains when the app is send into the background, and keeps the service active. For all emulated devices, I verified that the sync was able to run in the background for at least 1 minute, and also when the screen is turned off
- (Original metrics, before the 5 minute limit was added): On the Samsung Galaxy S23 Ultra device, I tested that the sync ran continuously for 50 minutes before opening the app again (testing using a large file system sync, which is very slow on Android, ideal for simulating a slow sync target). The logs showed that the sync did run continuously, though at a slower rate which was approximately 25% of the rate (1 item per 4 seconds) at which items are synced when the app is in foreground (1 item per second), when measured at the point just before bringing the app back to the foreground. When having battery optimisations for the app set to Optimised, the immediate rate of the sync was approximately 50% of the Unoptimised rate when in the background (1 item per 8 seconds). However, I also observed that a similar rate can be reached when the app is Unoptimised, when the service has been running for a long time with the app in the background
- Tested notification request access on first sync works (which waits for a response to the prompt, before sync goes ahead), and verified it does not prompt again when granted. The service triggers as soon as sync starts, so the app can be backgrounded any time from then
-  There may be a delay (up to a few seconds) in showing or hiding the notification on Android, or it may not show at all when the sync executes quickly. Testing and logging showed the foreground service execution is not dependent of whether the notification shows, but starting and stopping of the service happens instantly. This means there is no issue with a service not starting if another sync triggered straight after a service stops. When typing in a note, the notification can be seen appearing and disappearing in the status bar
- Verified the icon and text on the notification is correct. On Android 13 or below, there is no prompt to give permission, but a notification is displayed regardless
- If the user chooses 'do not allow notification' when clicking sync, they will be prompted one more time on the next sync, and if they choose do not allow that time, it will be remembered permanently, unless the user changes notification settings for the app, or clears app data
- On Android 16, I verified that notification and sync continues if I kill the app in recent apps (this behaviour may only be possible in some Android versions)

**Further testing on Samsung Galaxy S23 Ultra on Android 14:**

- Verified that manually cancelling the sync results in the notification being dismissed and the background task removed when referring to the recent app list (as shown on Android 14)
- Verified that if the app is already in background at the point when the sync has been running for approximately 5 minutes, that the notification is dismissed and the background task removed when referring to the recent app list. Additionally, if the sync was a full sync, the notification is dismissed when the upload step has completed and the background task removed when referring to the recent app list. When bringing the app back into the foreground, the sync continues where it left off (without being cancelled) but the notification does not come back. If cancelling the sync and retriggering it, then the notification will come back if there are outgoing changes to upload
- Repeated the test where the app is in foreground at the time the 5 minute limit is reached, or the delta step is reached for a full sync

**See video on Android 16, demonstrating the granting of notification access and syncing in the background:**

https://github.com/user-attachments/assets/25c73e9c-f303-4685-9516-31cefb42f2ac

**Screenshot of the new notification:**

Capture

### CODE REVIEW

File: packages/app-mobile/ios/Joplin/Info.plist [hunk 1]
```diff
@@ -116,5 +116,13 @@
...
 	<key>UIViewControllerBasedStatusBarAppearance</key>
 	<false/>
+	<key>UIBackgroundModes</key>
+	<array>
+		<string>processing</string>
+	</array>
+	<key>BGTaskSchedulerPermittedIdentifiers</key>
+	<array>
+		<string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
+	</array>
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/app-mobile/ios/Joplin/Info.plist]: Done

File: packages/app-mobile/package.json [hunk 2]
```diff
@@ -62,6 +62,7 @@
     "punycode": "2.3.1",
     "react": "19.1.4",
     "react-native": "0.81.6",
+    "react-native-background-actions": "4.1.0",
```
Comment (personalizedrefrigerator) [inline packages/app-mobile/package.json:65]: Note: [Based on this comment,](https://github.com/Rapsssito/react-native-background-actions/issues/249#issuecomment-4200176016) `react-native-background-actions` *might* not be as actively maintained as it used to be. It [does seem to be a dependency for a large number of other projects](https://github.com/Rapsssito/react-native-background-actions/network/dependents) though.

**Comment by personalizedrefrigerator** [comment]
> I would appreciate if someone could build and test this code on iOS to verify whether the solution works. But likely if there is an issue, it would just be regarding whether sufficient permissions have been allocated in packages/app-mobile/ios/Joplin/Info.plist

I've tested this on an iOS simulator:
- About 30 seconds after sending the app to the background during the initial sync, the app crashes with `EXC_BAD_ACCESS (Code=EXC_I386_GPFLT)` and XCode starts an LLDB session, paused on [this line](https://github.com/Rapsssito/react-native-background-actions/blob/3623c29cb6c23b9cb5860d1229e343e57a6490eb/ios/RNBackgroundActions.m#L34).
  - So far, I've observed this crash twice, both during the initial sync of just-created profiles.
- With the app in the foreground, Joplin Server sync seems to work.
  - My first sync attempt, I observed unexpected conflicts, but this might have been related to issues with my local setup.
  - I'm syncing on a new profile and didn't change any notes.
  - "Background service started" and "Background service stopped" messages are logged.

**Comment by mrjo118** [comment]
@personalizedrefrigerator Thank you for testing this!

> So far, I've observed this crash twice, both during the initial sync of just-created profiles.

Does this happen only after you have set up the sync for the profile, or does it happen before the sync has run as well?

Also, does the background sync ever work at all, or it crashes every time you attempt to sync a change?

**Comment by personalizedrefrigerator** [comment]
> Also, does the background sync ever work at all, or it crashes every time you attempt to sync a change?

I've only done limited testing, but so far, it seems to be only the first sync after setting up the profile. It doesn't seem to happen if no sync target is configured.

**Edit**: It also doesn't seem always to during the initial sync. Testing with an account that has only a single note (and E2EE disabled), the crash doesn't occur.

**Comment by mrjo118** [comment]
@personalizedrefrigerator I did some further research, and it looks like when using the beginBackgroundTaskWithName method of running a task in the background on iOS, the task must be completely stopped when iOS decides the time is up, otherwise it will cause the crash which you mentioned. The library supports an [expiry handler](https://www.npmjs.com/package/react-native-background-actions#expiration) for iOS. However, because of the way the sync is implemented in Joplin, I don't think there is a way to guarantee the sync can be stopped on the spot, as cancelling the sync would still continue the any upload / download operation in progress, before the sync would stop. It seems like around 30 seconds is a common point for which a beginBackgroundTaskWithName task would expire, even if theoretically it could last up to 2 or 3 minutes. The only other suitable method on iOS for achieving an overruning task seems to be the [URLSessionConfiguration.background method](https://developer.apple.com/documentation/foundation/urlsessionconfiguration/background(withidentifier:)), but React Native libraries which use this are scarce, and ChatGPT suggests that wouldn't work for in memory upload / download transfers, but they have to be file based so the job can be handed over to the OS.

As there is a real risk of creating instability on iOS with the react-native-background-actions implementation, and I don't think there is any other suitable solution for iOS, I feel that this would need to be an Android only solution. I will update the PR to make the implementation apply to Android only. I have not experienced any crashes on 4 different versions of Android though, so the Android solution seems to be solid, aside from performance degradation over time when in the background / when battery optimisations are not disabled.

File: packages/lib/registry.ts [hunk 3]
```diff
@@ -249,6 +285,48 @@ class Registry {
...
+		try {
+			return await service.start(async () => {
+				let response = null;
+				this.logger().debug(`registry.startSync [${uid}]: Background service started`);
+				response = await sync.start(options);
+				// Stop the service explicitly, as on some devices such as Samsung phones, the notification for the foreground service gets frozen while the
+				// app is in the background, preventing the notification being updated via updateNotification and preventing it dismissing automatically
+				await service.stop();
+				this.logger().debug(`registry.startSync [${uid}]: Background service stopped`);
+				return response;
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: Done

File: packages/lib/registry.ts [hunk 4]
```diff
@@ -4,6 +4,37 @@ import shim from './shim';
...
+	} | undefined;
+	foregroundServiceType?: ('dataSync' | 'mediaPlayback' | 'phoneCall' | 'location' | 'connectedDevice' | 'mediaProjection' | 'camera' | 'microphone' | 'health' | 'remoteMessaging' | 'systemExempted' | 'shortService' | 'specialUse')[] | undefined;
+}
+
+export interface BackgroundService {
+	start(
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Assigning types to these variables would be too big of a refactoring
+		task: (taskData?: any)=> Promise<void>,
+		options: BackgroundServiceOptions
+	): Promise<void>;
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: The return type of `Promise<void>` is correct, I have updated the startSync function appropriately

File: packages/lib/registry.ts [hunk 5]
```diff
@@ -209,7 +244,7 @@ class Registry {
 									Setting.setValue(contextKey, JSON.stringify(newContext));
 								};
 							}
-							newContext = await sync.start(options);
+							newContext = await this.startSync(sync, options);
 							Setting.setValue(contextKey, JSON.stringify(newContext));
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: Good spot, that was indeed a significant issue. I have now remedied this

**Comment by mrjo118** [comment]
Unfortunately after further testing I found a couple of issues, when doing very long syncs:
1. I experienced no issues uploading many changes over a long period, and no pauses in the sync execution, even if it was severely throttled. However, testing on my Samsung S23 with battery optimisations disabled, I found that when the sync included the delta step, after as little as 10-11 minutes, it is possible for the execution to stall and no longer continue, but keep the foreground service and notification open. It literally stopped on the first iteration of the delta step, just after reporting it was processing the first record, so it seems to be due to a different type of execution process running, rather than a random point at which it stalled. I believe that this is likely due to CPU starvation due to the OS throttling the background activity, and preventing network calls from successfully completing. I did not have this issue when doing a full sync, if the sync ran for a much shorted period of a few minutes. This could potentially leave the notification open for hours, and still drain battery while the foreground service is active
2. In an attempt to work around this, I tried locally altering the TaskQueue waitForResult function to timeout and throw an exception after 60 seconds. After building and running this code, I triggered a large sync and left it to run overnight. In the morning, I woke up to find that the notification was gone and the sync looked to be complete. However, when I checked the logs, it turned out that even after the sync had finished, further syncs were still able to be triggered while the app was in the background. The log showed the sync had worked through all the fetches at a rate of 2-4 seconds while in the background, but after the sync had completed, it triggered about an hour later, and then again about 3 hours later. When my alarm rang and I'd started interacting with my phone, the Joplin app had also woken up and started syncing in the background, without showing any notification. So what this seems to show is it is possible for there to be a process leak using this library when the process is starved of sufficient resources to continue, causing the foreground service to continue running when the stop signal should have been sent. I'm surprised that it was still able to complete the sync with multiple attempts however, under those circumstances.

In conclusion, it seems that background sync on Android using a foreground service does work, and works well for a short enough period, but if it runs for a long time, it can become unstable, due to heavy activity throttled by the OS, likely because it is not running in the native layer.

@laurent22 Do you think this Android implementation would be suitable if we set a time limit for the sync to auto cancel if it has been running in the background for a maximum amount of time? Maybe 5 minutes? When cancelling, it would still have to wait for the current download / upload to complete before the sync would cancel and cause the foreground service to stop, but I wouldn't expect it to ever take minutes for that to finish off.

EDIT: I have updated the PR and description with this change, but instead of cancelling the sync, I just stop the service only. In testing I've found that this solution works well

File: packages/lib/registry.ts [hunk 6]
```diff
@@ -249,6 +286,71 @@ class Registry {
...
+
+		if (sync.state() !== 'idle') return null;
+		options = { ...options, appIsActive: service ? service.appIsActive : null };
+		const uid = uuid.create();
+
+		try {
+			await service.start(async () => {
+				this.logger().debug(`registry.startSync [${uid}]: Background service started`);
+				this.syncStartTime = Date.now();
+				try {
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: This is acceptable behaviour for a simpler implementation. The assumption is incremental syncs should normally complete well within 5 minutes, so starting the timer when the sync starts is fine

File: packages/lib/registry.ts [hunk 7]
```diff
@@ -249,6 +286,71 @@ class Registry {
...
+		const service = this.backgroundService_;
+		if (!service) return;
+
+		const maxRuntimeExceeded = Date.now() - this.syncStartTime >= 5 * 60 * 1000;
+
+		if (service.isRunning() && !service.appIsActive() && maxRuntimeExceeded) {
+			try {
+				void service.stop();
+				this.logger().debug('registry.startSync: Background service stopped due to timing out');
+			} catch (e) {
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: This is intended. The cancelling() function does not perform a cancel but is just used to check whether a cancel should be performed. As it is in all the exact same conditions where the check needs to be applied, that the check whether the background service needs to be stopped is also applicable, it makes sense to perform this call within the cancelling() function itself. It is also intended that when the service is stopped, that the sync itself would not be stopped, as I want the sync to resume where it left off when the app is returned to the foreground, as per existing behaviour before background sync was implemented

[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: Ok I've removed the redundant try catch

File: packages/lib/registry.ts [hunk 8]
```diff
@@ -209,8 +245,8 @@ class Registry {
 									Setting.setValue(contextKey, JSON.stringify(newContext));
 								};
 							}
-							newContext = await sync.start(options);
-							Setting.setValue(contextKey, JSON.stringify(newContext));
+							newContext = await this.startSync(sync, options, contextKey);
+							if (newContext) Setting.setValue(contextKey, JSON.stringify(newContext));
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts:257]: This has now been addressed. For places which require awaiting scheduleSync, these will no longer run the sync using the service, and therefore the promise will resolve when the sync has actually completed. However the main sync paths including manual sync by the user, and automatic syncing as the user types, are not awaited, which means the service will be used in these cases on Android

File: packages/lib/Synchronizer.ts [hunk 9]
```diff
@@ -401,6 +403,7 @@ export default class Synchronizer {
 		}
 
 		this.state_ = 'in_progress';
+		this.syncStartTime_ = Date.now();
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/Synchronizer.ts]: I have now addressed this

[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/Synchronizer.ts]: I've fixed that other issue as well

File: packages/lib/registry.ts [hunk 10]
```diff
@@ -253,6 +293,66 @@ class Registry {
...
+
+	public stopBackgroundServiceIfExpired(syncStartTime: number) {
+		const service = this.backgroundService_;
+		if (!service) return;
+
+		const maxRuntimeExceeded = Date.now() - syncStartTime >= 5 * 60 * 1000;
+
+		if (service.isRunning() && !service.appIsActive() && maxRuntimeExceeded) {
+			void service.stop();
+			this.logger().debug('registry.sync: Background service stopped due to timing out');
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts]: I don't care about catching this error

File: packages/app-mobile/utils/buildStartupTasks.ts [hunk 11]
```diff
@@ -428,6 +430,36 @@ const buildStartupTasks = (
...
+					requestPermissions: async () => {
+						const response = await checkPermissions(PermissionsAndroid.PERMISSIONS.POST_NOTIFICATIONS);
+
+						// The POST_NOTIFICATIONS permission isn't supported on Android API < 33.
+						// (If unsupported, returns NEVER_ASK_AGAIN).
+						// On earlier releases, notifications should work without this permission.
+						if (response === PermissionsAndroid.RESULTS.DENIED) {
+							logger.warn('POST_NOTIFICATIONS permission was not granted');
+							return;
+						}
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/app-mobile/utils/buildStartupTasks.ts:454]: This is copied from existing behaviour for notifications for alarms. I don't intend to address an existing issue there if there is one

File: packages/lib/registry.ts [hunk 12]
```diff
@@ -253,6 +293,58 @@ class Registry {
...
 
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Assigning types to these variables would be too big of a refactoring
+	private async startSync(sync: Synchronizer, options: any, contextKey: string) {
+		// Service will only be populated for the native Android app
+		const service = this.backgroundService_;
+		if (!service || options.useService === false || service.isRunning() || !service.appIsActive()) return sync.start(options);
+
+		await service.requestPermissions();
+
+		if (sync.state() !== 'idle') return null;
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts:304]: As granting permission waits for user interaction the first time, the status check followed by starting the sync should only be done after the user interaction is complete, if somehow the sync got triggered via a different caller without the service and the sync is already running when the user makes their choice

File: packages/lib/registry.ts [hunk 13]
```diff
@@ -253,6 +293,58 @@ class Registry {
...
+				},
+				foregroundServiceType: ['dataSync'],
+			});
+		} catch (e) {
+			// If service.start is called while the service is already running, it will kill the existing service and start a new one. However if service.start
+			// is called while the app is in background, it will throw an exception and end up here. This can happen when the sync is triggered at the user
+			// configured sync interval. We should still run the sync normally in this case, in case there are other occasions this error will happen, but we
+			// it doesn't really matter if this sync does not complete while in the background, due to there being no service to keep it alive
+			this.logger().info('registry.sync: Starting background service failed, running sync directly', e);
+			return sync.start(options);
```
[the remark this answers is not in the document]
Comment (mrjo118) [reply packages/lib/registry.ts:343]: Now addressed by catching all exceptions within the task executed by service.start

**Comment by mrjo118** [comment]
Unfortunately, I am going to have to close this PR. While the current implementation does work, it is somewhat limited, because there is only support for background syncing of outgoing changes, and there is also no iOS support. The reason for no iOS support is mentioned in my earlier comment, and the reason for not supporting background sync during the delta step (incoming changes) of the sync is because the delta step will often halt the execution of the JS runtime while the app is in the background.

This happens even with battery optimisations are disabled for the app, and is not directly tied to how long the app is running in the background, as I was able to produce this in as little as one minute of running in the background. Even when setting a timeout so that the foreground service would automatically stop after 5 minutes, this timeout logic (via Promise.race) does not fire when the app becomes stuck on the delta step in this way, meaning the notification can persist indefinitely, and it will waste battery (as soon as the app is brought back to the foreground after this has happened, the execution becomes unblocked and the foreground service stops immediately). The only feasible explanation I could find for this behaviour, is some kind of resource starvation by the OS while the app is in the background, which prevents an awaited function from proceeding, possibly creating some kind of deadlock (strangely, new calls executed on timers are still able to execute in some cases, which means the JS runtime may not be completely halted). This means that the only way to stop the foreground service in this situation would be to implement some kind of timeout in native code. However this still would not address the fact that the service could be running unnecessarily for up to 5 minutes to effectively do nothing.

Although I have not experienced any such halting of execution during the upload steps of the sync during testing, I cannot guarantee that this could not happen there also, maybe when uploading a large file, because I don't really know the exact cause. My best guess is that avoiding the possibility of this happening requires either lowering the complexity of the synchronisation code (so that it does not have deep layers of nesting), or the issue is to do with limitations of the JS runtime when running in the background, and therefore avoiding this blocking would require duplicating the relevant sync logic in native code. So essentially there would likely need to be significant changes made to the sync logic in Joplin, in order to make this work reliably. Therefore this is unfortunately another dead end, unless some significant rewrite is the intended path to take
